plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "vendor id overflows");
+		return 0x0;
+	}
+
+	return (guint16) id;
+}
+
+static gboolean
+fu_plugin_thunderbolt_is_host (GUdevDevice *device)
+{
+	g_autoptr(GUdevDevice) parent = NULL;
+	const gchar *name;
+
+	/* the (probably safe) assumption this code makes is
+	 * that the thunderbolt device which is a direct child
+	 * of the domain is the host controller device itself */
+	parent = g_udev_device_get_parent (device);
+	name = g_udev_device_get_name (parent);
+	if (name == NULL)
+		return FALSE;
+
+	return g_str_has_prefix (name, "domain");
+}
+
+static void
+fu_plugin_thunderbolt_add (FuPlugin *plugin, GUdevDevice *device)
+{
+	FuDevice *dev_tmp;
+	const gchar *name;
+	const gchar *uuid;
+	const gchar *vendor;
+	const gchar *version;
+	const gchar *devpath;
+	gboolean is_host;
+	guint16 did;
+	guint16 vid;
+	g_autofree gchar *id = NULL;
+	g_autofree gchar *vendor_id = NULL;
+	g_autofree gchar *device_id = NULL;
+	g_autoptr(FuDevice) dev = NULL;
+	g_autoptr(GError) error = NULL;
+
+	uuid = g_udev_device_get_sysfs_attr (device, "unique_id");
+	if (uuid == NULL) {
+		/* most likely the domain itself, ignore */
+		/* TODO: handle devices in safe-mode */
+		return;
+	}
+
+	devpath = g_udev_device_get_sysfs_path (device);
+
+	g_debug ("adding udev device: %s at %s", uuid, devpath);
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	dev_tmp = fu_plugin_cache_lookup (plugin, id);
+	if (dev_tmp != NULL) {
+		g_debug ("ignoring duplicate %s", id);
+		return;
+	}
+
+	vid = fu_plugin_thunderbolt_udev_get_id (device, "vendor", &error);
+	if (vid == 0x0) {
+		g_warning ("failed to get Vendor ID: %s", error->message);
+		return;
+	}
+	did = fu_plugin_thunderbolt_udev_get_id (device, "device", &error);
+	if (did == 0x0) {
+		g_warning ("failed to get Device ID: %s", error->message);
+		return;
+	}
+
+	is_host = fu_plugin_thunderbolt_is_host (device);
+	vendor_id = g_strdup_printf ("TBT:0x%04X", (guint) vid);
+	device_id = g_strdup_printf ("TBT-%04x%04x", (guint) vid, (guint) did);
+
+	dev = fu_device_new ();
+	fu_device_set_id (dev, uuid);
+
+	fu_device_set_metadata (dev, "sysfs-path", devpath);
+	name = g_udev_device_get_sysfs_attr (device, "device_name");
+	if (name != NULL) {
+		if (is_host) {
+			g_autofree gchar *pretty_name = NULL;
+			pretty_name = g_strdup_printf ("%s Thunderbolt Controller", name);
+			fu_device_set_name (dev, pretty_name);
+		} else {
+			fu_device_set_name (dev, name);
+		}
+	}
+
+	vendor = g_udev_device_get_sysfs_attr (device, "vendor_name");
+	if (vendor != NULL)
+		fu_device_set_vendor (dev, vendor);
+	fu_device_set_vendor_id (dev, vendor_id);
+	fu_device_add_guid (dev, device_id);
+	version = g_udev_device_get_sysfs_attr (device, "nvm_version");
+	if (version != NULL)
+		fu_device_set_version (dev, version);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_ALLOW_ONLINE);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_ALLOW_OFFLINE);
```

Comment:
I don't think the ONLINE/OFFLINE is for EFI/Live -- this is more indicating if the device can be updated in gnome-software directly or needs to use the offline systemd upgrade mode. I suppose it might make sense to auto set `ALLOW_OFFLINE` if `ALLOW_ONLINE` is used, as I can't think of a situation where a device update would work on the Live desktop, but not in the single-user systemd upgrade mode as the firmware is saved to /var/lib in this case.